Sprint 16

Aleksandr0406 opened the pull request:
1. Экран со списком категорий переписан на архитектуру MVVM  с реализацией байндингов через замыкания

2. Сделаны экраны онбординга с использованием UIPageViewController

Aleksandr0406 committed:
feat: добавить два экрана онбординга

Aleksandr0406 committed:
style: общие правки по стайлу

Aleksandr0406 committed:
refactor: переписать экран списка категории на архитектуру MVVM

класс NewCategoryViewController

Helvetii951 commented on Tracker/App&Scene/SceneDelegate.swift:
```diff
@@ -8,12 +8,20 @@
 import UIKit
 
 class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+    var isThisNotFirstLoad: Bool {
+        get { UserDefaults.standard.bool(forKey: "IsThisNotFirstLoad") }
+        set { UserDefaults.standard.set(newValue, forKey: "IsThisNotFirstLoad") }
+    }
```
Для хранения подобных флагов стоит создать отдельный класс. С ростом приложения настроек и флагов будет больше. Хранить их в одном месте будет очень удобно.

Aleksandr0406 replied:
создал отдельный класс OnboardingData и поместил в него статическое свойство isThisNotFirstLoad и к нему уже обращаюсь в SceneDelegate. 
Нормально делать это свойство static?

Helvetii951 commented on Tracker/App&Scene/SceneDelegate.swift:
```diff
@@ -8,12 +8,20 @@
         guard let scene = (scene as? UIWindowScene) else { return }
         let window = UIWindow(windowScene: scene)
-        window.rootViewController = TabBarController()
+        let onBoardingViewController = OnBoardingViewController(transitionStyle: .scroll, navigationOrientation: .horizontal, options: nil)
+        let tabBarController = TabBarController()
```
Создавать контроллеры заранее чуть накладно. Лучше это делать на месте принятия решения.
```swift
window.rootViewController = isThisNotFirstLoad
                                ? TabBarController()
                                : OnBoardingViewController(transitionStyle: .scroll,
                                                           navigationOrientation: .horizontal,
                                                           options: nil)
```

Helvetii951 commented on Tracker/NewCategoryViewController/NewCategoryViewController.swift:
```diff
@@ -10,8 +10,7 @@ import UIKit
     
-    private var index: Int = 0
-    private var names: [String] = []
+    private var viewModel: ViewModel = ViewModel()
```
Для гибкости и тестируемости кода желательно объявлять вью-модель протоколом и как-нибудь инжектить.

[Неплохая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/).

Aleksandr0406 replied:
статью смотрю, спасибо, пока оставил старое решение

Helvetii951 commented on Tracker/NewCategoryViewController/NewCategoryViewController.swift:
```diff
@@ -10,8 +10,7 @@ import UIKit
     
-    private var index: Int = 0
-    private var names: [String] = []
+    private var viewModel: ViewModel = ViewModel()
```
`ViewModel` слишком обобщенный нэйминг. Подойдет, например, `CategoriesViewModel`.

Helvetii951 commented on Tracker/NewCategoryViewController/Model/Model.swift:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+final class Model {
```
Так же и с `Model`.

Helvetii951 commented on Tracker/NewCategoryViewController/ViewModel/ViewModel.swift:
```diff
@@ -0,0 +1,25 @@
+    
+    private let model: Model = Model()
+    
+    func didCreateNewCategory(with newCategoryName: String) {
+        model.saveCreatedCategoryNames(with: newCategoryName)
+        guard let newCategoryNamesCreated = newCategoryNamesCreated else { return }
+        newCategoryNamesCreated()
+        }
```
Чуть "поехали" отступы. Общепринятые отступы помогают читать код быстрее. 
В общем случаее поможет выделить весь файл `cmd+A` и зажать `control+i`.

Helvetii951 replied:
Можно заменить на `newCategoryNamesCreated?()`.

Helvetii951 commented on Tracker/OnBoardingController/BlueOnBoardingViewController.swift:
```diff
@@ -0,0 +1,83 @@
+
+import UIKit
+
+final class BlueOnBoardingViewController: UIViewController {
```
`BlueOnBoardingViewController` и `RedOnBoardingViewController` не отличаются набором UI компонентов. Можно создать один обобщенный контроллер, в него передавать необходимые для отображения данные.

Helvetii951 commented on Tracker/OnBoardingController/OnBordingViewController.swift:
```diff
@@ -0,0 +1,86 @@
+import UIKit
+
+final class OnBoardingViewController: UIPageViewController, UIPageViewControllerDataSource, UIPageViewControllerDelegate {
+    lazy var pageControl: UIPageControl = {
+        let pageControl = UIPageControl()
+        pageControl.numberOfPages = pages.count
+        pageControl.currentPage = 0
+        
+        pageControl.currentPageIndicatorTintColor = .black
+        pageControl.pageIndicatorTintColor = .gray
+        
+        pageControl.translatesAutoresizingMaskIntoConstraints = false
+        return pageControl
+    }()
+    
+    lazy var pages: [UIViewController] = [BlueOnBoardingViewController(), RedOnBoardingViewController()]
```
Свойства можно сделать приватными. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

Helvetii951 commented on Tracker/OnBoardingController/RedOnBoardingViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        button.addTarget(self, action: #selector(didTapAddButton), for: .touchUpInside)
+    }
+    
+    @objc private func didTapAddButton() {
+        // Записываем в переменную isThisNotFirstLoad(лежит в SceneDelegate), что онбординг случился
+        UserDefaults.standard.set(true, forKey: "IsThisNotFirstLoad")
+        
+        let tabBarViewController = TabBarController()
+        tabBarViewController.modalPresentationStyle = .fullScreen
+        present(tabBarViewController, animated: false)
+    }
```
Лучше забрать у контроллера онбординга обязанность навигации и отдать. ее тому, кто создал контроллер(в нашем случае это `SceneDelegate`). Передаем событие нажатия кнопки через, например, замыкание, а `SceneDelegate` уже пусть сам решает что делать.

Aleksandr0406 replied:
а что будет делать SceneDelegate? не могу понять. Он же уже запустил какой то из экранов. Сам он презентовать новый экран не может вроде как. ?

Aleksandr0406 committed:
refactor: общий рефакторинг проекта

Изменил логику показа экрана онбординга(вместо двух вьюконтроллеров теперь один)

Aleksandr0406 committed:
feat: добавить возможность чека привычкой с прошедшей датой
